### DISCUSSION TITLE
How to have column security with single end-point?

### DISCUSSION THREAD

**Discussion by ozum** [supabase/supabase#11670] [Questions]
Hi,

In postgrest and supabase context, DB views seem the most appropriate way to force column-specific permissions. However, if I create different views for different user levels (normal, manager, admin, etc.), I will have different endpoints (`api/employee`, `api/employee-with-salary`, `api/employee-with-admin-columns`, etc.) which result in:

- REST clients need to know the application user's level and select endpoints based on the user's level.
- `supabase-js` or any other ORM also needs to select from DB views based on the application user's level.

As a result, clients would have application logic, which I don't want.

Another option is to create a DB function for each table with sensitive columns and handle all authorization stuff in that DB function. The downside is, I need to create lots of functions and clients use not-so-lovable RPC interfaces. In that case (RPC & DB functions), I prefer to develop the app from scratch using a backend framework instead of using supabase.

I looked at PostgreSQL "SELECT rules" whether I can redirect select queries into different views based on a condition, so I have a single named view/endpoint for a single table. However, "SELECT rules" seem not to support this use case.

I want to have a single-named view/endpoint having different columns for each user level. How can I achieve this with supabase?

P.S. I posted a similar question to postgrest repo too, but supabase may have a specific solution to this issue.

**Comment by ozum** [comment]
At postGREST repo, @steve-chavez pointed [hybrid user group roles](https://postgrest.org/en/stable/auth.html#hybrid-user-group-roles). In brief, column-level grants can be used to limit column-level access. AFAIK, supabase JS client and supabase auth only support fixed database users.

Can custom database users be used with supabase JS client and supabase auth without being hacky?

**Reply by GaryAustin1** [reply]
You had to mint your jwt tokens to have other user roles.  But I don't believe with the new security changes you can do that now (you can't create a role for PostgREST to use now without super admin privilege).
And it was hacky also...

I think rpc or a view with the original table locked by RLS is the only way to go database side.

Or split out the columns to another table and then use RLS and clientside joins to bring it in when allowed.

**Reply by ozum** [reply]
Thanks, @GaryAustin1. I'm OK with views, but as I indicated in the original post, I need to have a single end-point. Do you have any suggestions to have a single end-point using views?

**Comment by ruggi99** [comment]
Hi, I think you can use `user_metadata` field of a single user.
You can set a JSON object that will be included in the JWT Token.
Metadata can be extracted during RLS policy evaluation.
And also you can decode the JWT Token client side to know which endpoint you can call and so it.
Requires duplicated logic (server and client side) but it shoul work

**Reply by ozum** [reply]
@ruggi99, thanks for the suggestion. It's a nice idea and could be used to provide clues to clients from the server.

For this use case, it is hard to implement, because there would be some major drawbacks:

* Duplicated logic (as you indicated)
* JWT becomes bigger and bigger until no more viable. There could be hundreds or thousands of endpoints.
* Most of the generic tools cannot be used with this approach, because clients must know that they must decode JWT and use those endpoints.

**Reply by ruggi99** [reply]
I don't quite understand your points.
- Duplicated logic could be resolved by creating a single rpc function that check user's metadata. It's not easy to always edit the function but it resolves the point. Also you can use a plenty of programming languages.
- Obviously you would not store the endpoints in the JWT Token, it would be problematic. Also because if you change an endpoint you have to renew the JWT Token. Going with RPC server side resolves it.
- Also this is resolved with RPC server side.

After all, I don't think Supabase is interested in altering current roles, as it would be a significant change and it'll require testing on various platforms. To workaround it they implemented metadata.
Also metadata is meant to store public data and not sensitive data.

EDIT: If Supabase had custom roles you would need a logic to calculate your endpoints and Supabase can't help you much in this. So custom roles from Supabase or metadata in users table are quite the same at my eyes

**Reply by ozum** [reply]
I may misunderstand your solution. I will try to explain the points I am stuck to currently. Could you please elaborate?

From your first message:
> And also you can decode the JWT Token client side to know which endpoint you can call and so it.

From your second message:
> Obviously you would not store the endpoints in the JWT Token

From the first message, I thought I should write endpoints to JWT. So the client reads the correct endpoints from there. The second message showed you didn't mean that. Without using RPC, how does the client know which view to call? "employee", "employee_with_salary_data", "employee_with_approve_process_data", "employee....", plus the other 15 views for the employee table?

> Duplicated logic could be resolved by creating a single rpc function that check user's metadata. It's not easy to always edit the function but it resolves the point. Also you can use a plenty of programming languages.

Do you mean this flow: RPC -> database function -> view -> table

For example, the Client calls `get_emplyee` RPC, database function `get_employee` decides the view to use based on metadata and selects from the view, and the view`employee_with_salary_data` selects data from table `private.employee`.

I evaluated this solution before, but database functions have to be defined with a return type, and views have different structures, with different numbers of columns. I can define the return type as the union of all column names of all views and provide null values for the missing values:

```sql

-- Database Function: get_employees()
...
IF level = 'admin' THEN
  RETURN QUERY SELECT a, b, c, d, e, f, g, NULL, x, y, NULL, z FROM private.employee;
ELSIF level = 'user' THEN
  RETURN QUERY SELECT a, b, NULL, NULL, NULL, NULL, NULL, NULL, x, NULL, NULL, NULL FROM employee_for_user;
ELSIF level = 'manager' THEN
  RETURN QUERY SELECT a, b, c, d, NULL, NULL, NULL, NULL, NULL, NULL, NULL, z FROM employee_with_salary_data;
ELSIF level = '...' THEN
  ....
ELSIF level = '...' THEN
ELSIF level = '...' THEN
END IF
```

I also need to write those functions for INSERT and UPDATE too... I need to sync all of the queries in those functions when the table is changed. This is very hard to maintain in the long run.

> After all, I don't think Supabase is interested in altering current roles

Maybe, but I opened a feature request #11745, because they mentioned custom roles are planned long before...

>  If Supabase had custom roles you would need a logic to calculate your endpoints and Supabase can't help you much in this.

If Supabase supports custom roles, I can use PostgreSQL grants to limit column access.

```sql
grant select (a, b, c, d, e, f, g, x, y, z) on employee to admin
grant select (a, b, x) on employee to user
grant select (a, b, c, d, z) on employee to manager
...
```

Now, I can set the app user to a custom database role when she is logged in, based on their level, and use a single "employee" endpoint for all user levels since PostgreSQL limits their access to columns.

`SET LOCAL ROLE manager`

Thanks again. You are taking your time for helping me.

**Reply by ruggi99** [reply]
Yeah, re-reading my first message it's not clear what I meant.
In my first message I meant to insert the role in the metadata, not endpoints.
Also I thought that `endpoints` meant a custom logic server side implemented by you that talks to database with `service_role` key.
This actually might be a solution. Server side code that bypasses RLS policies and fetches data from database as you want.

I think creating multiple views is not the right way.

One solution might be the reverse. Create multiple tables for different level access with foreign keys to a "primary" table (that is public).
Then create a single view with alle the data with joins, etc.
Enable RLS policies on the tables, reading the role from the metadata and JWT Token.
I don't know if it works, but I think (and hope) views will respect RLS policies returning void data.
EDIT: You can avoid creating the giant view, you can make multiple calls to the tables (RLS will work better I think and requires no logic to decide which endpoint call) and merge data client side.

I know that those 2 solutions are not ideal in terms of quality of code and maintenance but it's a personal decision of course.
Also Supabase it's not meant to give so fine-grained ACL (al least I think).

As you said, enabling custom roles would be ideal, but it's not trivial because requires changing postgres-meta, postgresql schema, frontend, docs and so on. Also I think there would be problems with PostgREST.

Let me know if it works the second solution

**Reply by ozum** [reply]
> This actually might be a solution. Server side code that bypasses RLS policies and fetches data from database as you want.

This works, and I'm using it with Prisma on the server side for custom endpoints as you described. For non-custom endpoints, I just don't want to write repetitive server-side code for each table.

> One solution might be the reverse. Create multiple tables for different level access with foreign keys to a "primary" table (that is public).

Yes, this is also a very good solution if the roles do not have intersecting requirements. For example, I only need two levels for `company`. I created two tables `company_public` and `company_private`. An admin can use both whereas normal users can only access `company_public`. This solution also reduces database disk access significantly, because `company_public` is accessed 1000 times more than `company_private` because data in `company_private` is rarely used. 

Unfortunately, for other tables such as `employee`, this doesn't work. I can't divide columns for tables, because there are intersections between roles.

|column | admin | user | manager |
|---|---|---|---|
| `a` | x | x | x |
| `c` | x | | x |
| `x` | x | x | |

> custom roles would be ideal, but it's not trivial because requires changing postgres-meta, postgresql schema, frontend, docs and so on.

Yes, it is not trivial. According to [this](https://github.com/supabase/supabase/discussions/3107#discussioncomment-1292883) and [this](https://github.com/supabase/supabase/discussions/3232#discussioncomment-1343550), it has been started 1.5 years ago. It might not be prioritized at that time, but I hope it will be.

> Also I think there would be problems with PostgREST.

PostgREST has support for custom roles. Please see [Hybrid User-Group Roles](https://postgrest.org/en/stable/auth.html#hybrid-user-group-roles)

> Let me know if it works the second solution

Both solutions work for some tables very well. I also have lots of other types of tables for which I need a solution too.

**Reply by ruggi99** [reply]
Mmm ok, I understand you problem, but if you split your tables based on groups I think you can do it.
So try to not split tables based on roles, but on groups (like on Linux for example).
Then a group as a set of permissions that can intersect with each other. Roles can be part of multiple groups and vice-versa.
Does it work?

**Reply by ozum** [reply]
Technically it is possible but is really hard,

Same simple example as before,

|column | admin | user | manager |
|---|---|---|---|
| `a` | x | x | x |
| `c` | x | | x |
| `x` | x | x | |

I need to split it into 3 tables: "table_with_a", "table_with_c", "table_with_x".

|Group|Access|
|---|---|
|Group 1| table_with_a, table_with_c|
|Group 2| table_with_a, table_with_x|

|Member|Groups|
|---|---|
|admin|Group 1, Group 2|
|user|Group 2|
|manager|Group 1|

Assume tens of tables with tens of columns. My mental health does not maintain this :)

**Reply by ruggi99** [reply]
Last solution is to use Supabase Edge Functions and handle all logic there, server side, using the service_role key and users' metadata.
Of course is worse than custom roles, but can it work in the meantime?

**Reply by ozum** [reply]
Yes, I'm already using it with Prisma on the server side for some endpoints. IMHO, overusing `service_role` destroys the whole purpose of supabase. If I'm going to develop all by myself, then I would choose hapi, fastify or nest as a backend and develop without any limitation of the supabase.

**Reply by JulianSwales** [reply]
I had a case where I required different record return data from one function (dynamic pivot data) and ran into the data type issue. I worked around it by returning a setof JSON and in my function where I do all the logic I return next row_to_json(r) 
This put all my logic in one place in a single function whilst allowing a dynamic data return. Not sure if this would work for you..

**Reply by ozum** [reply]
@JulianSwales, thanks. Nice idea for terminating queries (where results are meant to be used in a client), and I will use it for some of my queries. However, for the remaining, I need to use full SQL on returned results and also join the results of the function with other tables too.

**Comment by techandmedia** [comment]
I don't know if this related to your question but I can do multiple queries with a single endpoint

```js
// From front end, just pass the params as you see fit (based on the user role)
const PARAMS: ISupabase = {
  url: '/api/supabase',
  // Schema public sudah default, tidak perlu define
  data: {
    table: 'item',
    // select: '*',
    select: 'item_code, item_name',
    page: 1,
    size: 5,
  },
  // data: {
  //   schema: 'cst',
  //   table: 'address',
  //   select: '*',
  //   page: 1,
  //   size: 5,
  // },
  // data: {
  //   schema: 'ksk',
  //   table: 'v_cart', // getting data from a 'view' table
  //   select: '*',
  //   page: 1,
  //   size: 10,
  // },
};
```

```js
// I am using next js api so should be easy to adjust to your 'backend'
const { token, schema, table, select, page, size } = req.body;

const options = {
  schema: schema || 'public',
  global: {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  },
};
const supabase = supabaseClient(options);
const { count, error: errorCount } = await supabase
  .from(table)
  .select(select, { count: 'exact', head: true });
```
